contracts/utils/ExternalMulticall.sol

```diff
@@ -62,6 +62,59 @@ abstract contract ExternalMulticall is IExternalMulticall {
         }
     }
 
+    /// @notice Batches calls to external contracts and reverts as soon as one
+    /// of the batched calls reverts
+    /// @param targets Array of target addresses of batched calls
+    /// @param data Array of calldata of batched calls
+    /// @param values Array of values to be sent with each call
+    /// @return returndata Array of returndata of batched calls
+    function externalMulticallWithValue(
+        address[] calldata targets,
+        bytes[] calldata data,
+        uint256[] calldata values
+    ) public payable virtual override returns (bytes[] memory returndata) {
+        uint256 callCount = targets.length;
+        require(
+            callCount == data.length && callCount == values.length,
+            "Parameter length mismatch"
+        );
+        uint256 accumulatedValue = 0;
+        returndata = new bytes[](callCount);
+        for (uint256 ind = 0; ind < callCount; ) {
+            accumulatedValue += values[ind];
+            require(msg.value >= accumulatedValue, "Insufficient value");
+            require(
+                targets[ind].code.length > 0,
+                "Multicall target not contract"
+            );
```

Comment:
> I wonder if we should allow `targets[ind].code.length==0` if `data[ind].length==0` and `values[ind] != 0`. I'm thinking of a use-case where the user needs to multicall a `withdraw()` function that only transfers to `msg.sender` (though this would also mean that we would need a `receive()`).

Then this also means we would have to add a sweep function with some access control. Maybe we can implement an efficient solution like this; when it is called, `externalMulticallWithValue()` updates a storage variable `owner` to the `msg.sender`. And if some value comes back to `receive()`, it can send the value back to the `owner`. Also initialize `owner` back to `address(this)` in the end of the function.
This can also be done by just forwarding any value that comes to `receive()` to the `tx.origin`.

Reply:
> Then this also means we would have to add a sweep function with some access control

That's true, we either need `externalMulticallWithValue()` to not validate `values` based on `msg.value` or we need a sweep function. The latter is probably less error prone. 

> Maybe we can implement an efficient solution like this

That sounds very hacky. `externalMulticallWithValue()` has to be behind `onlyOwner` anyway so I don't see why we don't have the sweep function behind that as well.